src/Raven.Client/Documents/Session/Operations/StreamOperation.cs

```diff
@@ -580,24 +575,19 @@ private static void HandleStreamQueryStats(JsonOperationContext context, StreamR
 
             private void CheckIfContextNeedsToBeRenewed()
             {
-                if (_builderContext.CachedProperties.PropertiesDiscovered > 1024 ||
-                    _builderContext.AllocatedMemory > 4 * 1024 * 1024 ||
+                if (_builderContext.AllocatedMemory > 4 * 1024 * 1024 ||
                     _docsCountOnCachedRenewSession > _maxDocsCountOnCachedRenewSession)
                 {
                     _builderContext.Reset();
                     _builderContext.Renew();
                     _docsCountOnCachedRenewSession = 0;
                     return;
                 }
-                
-                if (_cachedItemsRenew)
-                {
-                    _builderContext.CachedProperties.Reset();
-                    _builderContext.CachedProperties.Renew();
 
-                    ++_docsCountOnCachedRenewSession;
-                    _cachedItemsRenew = false;
-                }
+                if (_builder.NeedResetPropertiesCache())
+                    _builderContext.CachedProperties = new CachedProperties(_builderContext);
```

Comment:
What are we doing with the old `CachedProperties`?

Reply:
I don't see `Dispose` for `CachedProperties` and as I understand it we do the same when we call context reset&renew with extra memory free (that we don't need here until we have high memory usage and we will reset&renew the whole context).
I mimicked the behavior from `UnmanagedJsonParserHelper.ReadArrayToMemory` and it seems to me to be more efficient.

Reply:
I don't see `Dispose` for `CachedProperties` and as I understand it we do the same when we call context reset&renew with extra memory free (that we don't need here until we have high memory usage and we will reset&renew the whole context).
I mimicked the behavior from `UnmanagedJsonParserHelper.ReadArrayToMemory` and it seems to me to be more efficient.

Reply:
I would rather reset the `CachedProeprties`, so we'll clear the values, but not allocate new memory.
That would reduce the amount of memory we allocate

Reply:
Reset the `CachedProeprties` doesn't clean all collections in `CachedProeprties` and particularly doesn't clean `_propertiesSortOrder` and `_propertyNameToId` that cause the problem.
I think it is on purpose so may I separate it to `SoftReset` and `HardReset` so the `HardReset` will clean it all?

Reply:
There are two modes here. Reset, where we remember the old properties and clear, where we don't.
The whole idea is that we can reduce memory allocations.